doc/design/concurrent_programming.md

```diff
@@ -0,0 +1,163 @@
+# Design Doc: Concurrent Programming with Fluid
+
+With PaddlePaddle Fluid, users describe a program other than a model.  The program is a [`ProgramDesc`](https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/framework/framework.proto) protobuf message. TensorFlow/MxNet/Caffe2 applications generate protobuf messages too, but their protobuf messages represent the model, a graph of operators, but not the program that trains/uses the model.   
+
+Many know that when we program TensorFlow, we can specify the device on which each operator runs.  This allows us to create a concurrent/parallel AI application.   An interesting questions is **how does a `ProgramDesc` represents a concurrent program?**  
+
+The answer relies on the fact that a `ProgramDesc` is similar to an abstract syntax tree (AST) that describes a program.  So users just program a concurrent program that they do with any concurrent programming language, e.g., [Go](https://golang.org).
+
+## An Analogy
+
+The following table compares concepts in Fluid and Go
+
+| Go | Fluid |
```

Comment:
The concepts of a programming language is not only for Go, all programming language have. People may not familiar with Go.

Reply:
Other languages don't have goroutine.

Reply:
I see. You can use the concept "coroutine". But never mind, either is OK.

Reply:
No, coroutine is completely different from goroutine. Nothing similar to each other indeed.

Reply:
goroutine is preemptive scheduled, coroutine need to give up the occupation of OS thread before it could run other coroutines. goroutine is implemented as a thread pool in the Go runtime library, coroutine is a special data type -- a stack of stack.